### PR TITLE
vim-patch:8.2.3520: cannot define a function for thesaurus completion

### PR DESCRIPTION
**Pull request by mcepl**
Problem:    Cannot define a function for thesaurus completion.
Solution:   Add 'thesaurusfunc'. (Yegappan Lakshmanan, closes vim/vim#8987,
            closes 8950)
https://github.com/vim/vim/commit/160e994d768d03a3c826b58115cde94df8fce607

### CODE REVIEW

**Review by zeertzjq — COMMENTED** [review]
~You forgot to add `BV_THSFU` to `option_defs.h`.~

**Review by seandewar — CHANGES_REQUESTED** [review]
- There's a few lint failures that `make lint` shouid've caught. https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#coding
- This doesn't compile as per @zeertzjq's comment.

**Comment by zeertzjq** [comment]
I think you also need to port patch 8.2.3525.

**Comment by mcepl** [comment]
all 8.2.3521, 8.2.3525, and 8.2.3528 should be included.

**Comment by seandewar** [comment]
This isn't compiling and lint still needs addressing

**Review by seandewar — COMMENTED** [review]
Style comments to help appease the linter

**Review by seandewar — COMMENTED** [review]
Many of these comments are probably nit-picks, but functionally I think this PR looks good.
We'll have to see what janlazo thinks :+1:

File: src/nvim/edit.c [hunk 1]
```diff
@@ -4096,7 +4111,14 @@ int ins_compl_add_tv(typval_T *const tv, const Direction dir, bool fast)
     return FAIL;
   }
   return ins_compl_add((char_u *)word, -1, NULL,
-                       (char_u **)cptext, true, &user_data, dir, flags, dup);
+                       (char_u * *)cptext, true, &user_data, dir, flags, dup);
+}
+
+/// Returns TRUE when using a user-defined function for thesaurus completion.
+static int thesaurus_func_complete(int type)
```
Comment (janlazo) [inline src/nvim/edit.c:4118]:
```
/// Returns true when using a user-defined function for thesaurus completion.
static bool thesaurus_func_complete(int type)
```

**Review by janlazo — APPROVED** [review]
LGTM, ignoring style changes.
Fix the code style, address  `thesaurus_func_complete()`, and I'll merge.

File: src/nvim/edit.c (hunk 1 above)
Comment (janlazo) [reply src/nvim/edit.c:4118]: `type` should be an enum/typedef.

Comment (mcepl) [reply src/nvim/edit.c:4118]: @janlazo Changing `int` to `bool` is simple, but should I create that `enum` or will we wait on the upstream?

File: src/nvim/edit.c [hunk 2]
```diff
@@ -4114,11 +4115,10 @@ int ins_compl_add_tv(typval_T *const tv, const Direction dir, bool fast)
 }
 
 /// Returns TRUE when using a user-defined function for thesaurus completion.
```
Comment (zeertzjq) [inline src/nvim/edit.c:4117]: `TRUE` -> `true`

File: src/nvim/edit.c (hunk 1 above)
Comment (seandewar) [reply src/nvim/edit.c]: Create it yourself. You can do this in a future PR.

File: src/nvim/edit.c [hunk 3]
```diff
@@ -2899,7 +2900,7 @@ static void ins_compl_dictionaries(char_u *dict_start, char_u *pat, int flags, i
   char_u *ptr;
   char_u *buf;
   regmatch_T regmatch;
-  char_u **files;
+  char_u * *files;
```
Comment (seandewar) [inline src/nvim/edit.c:2903]:
```suggestion
  char_u **files;
```

These still need changing.

File: src/nvim/edit.c (hunk 2 above)
Comment (mcepl) [reply src/nvim/edit.c:4117]: When I do `src/clint.py src/nvim/edit.c` I get 294 errors, should I fix all of those? I don't think it has any relation to this vim patch.

Yes, this `TRUE` is changed.

File: src/nvim/edit.c (hunk 1 above)
Comment (mcepl) [reply src/nvim/edit.c]: @seandewar BTW, I am not the author of the original vim patch. But I guess it means, that this conversation can be marked as resolved, right?

File: src/nvim/edit.c (hunk 3 above)
Comment (mcepl) [reply src/nvim/edit.c:2903]: Done.

File: src/nvim/edit.c (hunk 2 above)
Comment (zeertzjq) [reply src/nvim/edit.c]: > should I fix all of those?

I think you just need to pass lint CI.

Comment (mcepl) [reply src/nvim/edit.c]: That I have been passing for some time already.

File: src/nvim/edit.c (hunk 3 above)
Comment (seandewar) [reply src/nvim/edit.c]: The others also need fixing.

File: src/nvim/edit.c [hunk 4]
```diff
@@ -2544,7 +2545,7 @@ static void ins_compl_longest_match(compl_T *match)
  * Add an array of matches to the list of matches.
  * Frees matches[].
  */
-static void ins_compl_add_matches(int num_matches, char_u **matches, int icase)
+static void ins_compl_add_matches(int num_matches, char_u * *matches, int icase)
```
Comment (seandewar) [inline src/nvim/edit.c:2548]:
```suggestion
static void ins_compl_add_matches(int num_matches, char_u **matches, int icase)
```

File: src/nvim/edit.c [hunk 5]
```diff
@@ -2990,7 +2991,7 @@ static void ins_compl_dictionaries(char_u *dict_start, char_u *pat, int flags, i
   xfree(buf);
 }
 
-static void ins_compl_files(int count, char_u **files, int thesaurus, int flags,
+static void ins_compl_files(int count, char_u * *files, int thesaurus, int flags,
```
Comment (seandewar) [inline src/nvim/edit.c:2994]:
```suggestion
static void ins_compl_files(int count, char_u **files, int thesaurus, int flags,
```

File: src/nvim/edit.c [hunk 6]
```diff
@@ -4096,7 +4111,14 @@ int ins_compl_add_tv(typval_T *const tv, const Direction dir, bool fast)
     return FAIL;
   }
   return ins_compl_add((char_u *)word, -1, NULL,
-                       (char_u **)cptext, true, &user_data, dir, flags, dup);
+                       (char_u * *)cptext, true, &user_data, dir, flags, dup);
```
Comment (seandewar) [inline src/nvim/edit.c:4114]:
```suggestion
                       (char_u **)cptext, true, &user_data, dir, flags, dup);
```

File: src/nvim/edit.c [hunk 7]
```diff
@@ -4116,7 +4138,7 @@ static int ins_compl_get_exp(pos_T *ini)
   static buf_T *ins_buf = NULL;          // buffer being scanned
 
   pos_T *pos;
-  char_u **matches;
+  char_u * *matches;
```
Comment (seandewar) [inline src/nvim/edit.c:4141]:
```suggestion
  char_u **matches;
```

Comment (mcepl) [reply src/nvim/edit.c:4141]: That's weird. All these four changes seem to the best of my knowledge applied to the current PR.

**Review by janlazo — APPROVED** [review]

**Comment by mcepl** [comment]
Thank you very much, @janlazo.

**Comment by janlazo** [comment]
@mcepl  Thanks for your contribution. 
@seandewar @zeertzjq  Thanks for your continued assistance and your attention to detail. You both made it a lot easier to review this MR.

**Comment by David-Else** [comment]
How can I use this in my `init.lua`? I used to have in `init.vim`:
```
func Thesaur(findstart, base)
    if a:findstart
	let line = getline('.')
	let start = col('.') - 1
	while start > 0 && line[start - 1] =~ '\a'
	   let start -= 1
	endwhile
	return start
    else
	let res = []
	let h = ''
	for l in split(system('aiksaurus '.shellescape(a:base)), '\n')
	    if l[:3] == '=== '
	    	let h = substitute(l[4:], ' =*$', '', '')
	    elseif l[0] =~ '\a'
		call extend(res, map(split(l, ', '), {_, val -> {'word': val, 'menu': '('.h.')'}}))
	    endif
	endfor
	return res
    endif
endfunc

set thesaurusfunc=Thesaur   " use the 'thesaurusfunc' for external thesaurus
```
But I can't seem to work out the synax for `vim.opt.thesaurusfunc` in Neovim 0.7. Cheers!

**Comment by mcepl** [comment]
I have this:
```vim
func Thesaur(findstart, base)
    if a:findstart
        return searchpos('\<', 'bnW', line('.'))[1] - 1
    endif
    let res = []
    let h = ''
    for l in systemlist('aiksaurus '.shellescape(a:base))
        if l[:3] == '=== '
        let h = '('.substitute(l[4:], ' =*$', ')', '')
        elseif l ==# 'Alphabetically similar known words are: '
        let h = "\U0001f52e"
        elseif l[0] =~ '\a' || (h ==# "\U0001f52e" && l[0] ==# "\t")
        call extend(res, map(split(substitute(l, '^\t', '', ''), ', '),
        \                    {_, val -> {'word': val, 'menu': h}}))
        endif
    endfor
    return res
endfunc

set thesaurusfunc=Thesaur
```

**Comment by David-Else** [comment]
@mcepl Can you re-create that in an `init.lua` file, how can I define a vimscript user function and reference it?

**Comment by mcepl** [comment]
I probably could, but it is just a waste of time. Put this into `~/.config/nvim/plugin/mythesaurus.vim` and be done with it. If you can persuade me with reasonable arguments this has any disadvantages against your Lua solution, I may try to rewrite it.

**Comment by David-Else** [comment]
@mcepl I did the following and it worked, thanks for answering. Lua solution is optimal, but I can't be bothered myself :)

```
vim.cmd [[
func Thesaur(findstart, base)
    if a:findstart
	let line = getline('.')
	let start = col('.') - 1
	while start > 0 && line[start - 1] =~ '\a'
	   let start -= 1
	endwhile
	return start
    else
	let res = []
	let h = ''
	for l in split(system('aiksaurus '.shellescape(a:base)), '\n')
	    if l[:3] == '=== '
	    	let h = substitute(l[4:], ' =*$', '', '')
	    elseif l[0] =~ '\a'
		call extend(res, map(split(l, ', '), {_, val -> {'word': val, 'menu': '('.h.')'}}))
	    endif
	endfor
	return res
    endif
endfunc
]]

vim.opt.thesaurusfunc = 'Thesaur'
```

**Pull request merged**